Tests/Functional/Sources/ErrorFormattingTest.m

```diff
@@ -0,0 +1,130 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "BaseIntegrationTest.h"
+
+#import "GREYError.h"
+#import "EarlGrey.h"
+#import "FailureHandler.h"
+
+# pragma mark - GREYErrorFormatTestingFailureHandler
+
+/**
+ * Failure handler used for testing for correct formatting of errors
+ * made with GREYErrorFormatter
+ */
+@interface ErrorFormatTestingFailureHandler : NSObject <GREYFailureHandler>
+@property NSString *fileName;
+@property(assign) NSUInteger lineNumber;
+@property GREYFrameworkException *exception;
+@property NSString *details;
+@end
+
+@implementation ErrorFormatTestingFailureHandler
+
+- (void)handleException:(GREYFrameworkException *)exception details:(NSString *)details {
+  self.exception = exception;
+  self.details = details;
+}
+
+- (void)setInvocationFile:(NSString *)fileName andInvocationLine:(NSUInteger)lineNumber {
+  self.fileName = fileName;
+  self.lineNumber = lineNumber;
+}
+
+@end
+
+# pragma mark - ErrorFormattingTest
+
+/**
+ * Tests that the user-facing console output follows expectations.
+ */
+@interface ErrorFormattingTest : BaseIntegrationTest
+
+@end
+
+@implementation ErrorFormattingTest
+
+- (void)testElementNotFoundErrorDescription {
+  ErrorFormatTestingFailureHandler *handler = [[ErrorFormatTestingFailureHandler alloc] init];
+  [NSThread mainThread].threadDictionary[GREYFailureHandlerKey] = handler;
+  
+  id<GREYMatcher> matcher =
+      [[GREYHostApplicationDistantObject sharedInstance] matcherForFirstElement];
+  [[EarlGrey selectElementWithMatcher:grey_allOf(grey_kindOfClass([UITableViewCell class]),
+                                                 matcher, nil)]
+      performAction:grey_tap()];
+  [[EarlGrey selectElementWithMatcher:grey_text(@"Basic Views")]
+                    assertWithMatcher:grey_notNil()
+                                error:nil];
+  
+  NSString *expectedDetails1 = @"Interaction cannot continue because the desired element was not "
+                               @"found.\n"
+                               @"\n"
+                               @"Check if the element exists in the UI hierarchy printed below. If "
+                               @"it exists, adjust the matcher so that it accurately matches "
+                               @"the element.\n"
+                               @"\n"
+                               @"Element Matcher:\n"
+                               @"((kindOfClass('UILabel') || kindOfClass('UITextField') || "
+                               @"kindOfClass('UITextView')) && hasText('Basic Views'))\n"
+                               @"\n"
+                               @"App-side Screenshot at Point-of-Failure:";
+  NSString *expectedDetails2 = @"UI Hierarchy (ordered by window level, back to front):";
+  NSString *expectedDetails3 = @"Stack Trace: (";
+  for (NSString *expectation in @[expectedDetails1, expectedDetails2, expectedDetails3]) {
```

Comment:
Expand this. In the case of failure, the current format doesn't give you useful information, just saying "[handler.details containsString:expectation] is false".

Same below.

Reply:
Good point.